src/azul/portal_service.py

```diff
@@ -0,0 +1,154 @@
+                else:
+                    try:
+                        db = self._read_db(version)
+                    except NoSuchObjectVersion:
+                        # Wait for latest version to appear in S3.
+                        continue
+                db = operation(db)
+                if db is None:
+                    # Operation is read-only; we're done.
+                    break
+                else:
+                    self._write_db(db, version)
+            except VersionConflict:
+                # Retry with up-to-date DB
+                continue
+            else:
+                break
+
+    def _create_db(self) -> Tuple[JSONs, str]:
+        """
+        Write hardcoded portal integrations DB to S3.
+        :return: Newly created DB and accompanying version.
+        """
+        plugin = Plugin.load()
+        db = self.demultiplex(plugin.portal_db())
+        version = self._write_db(db, None)
+        return db, version
+
+    def _read_db(self, version: str) -> JSONs:
+        """
+        Retrieve specified version of portal DB from S3.
+        Raises `NoSuchObjectVersion` if the version is not found.
+        """
+        try:
+            response = self.client.get_object(Bucket=config.portal_db_bucket_name,
+                                              Key=config.portal_db_object_key,
+                                              VersionId=version)
+        except self.client.exceptions.NoSuchKey:
+            raise NoSuchObjectVersion(version)
+        json_bytes = response['Body'].read()
+        return json.loads(json_bytes.decode())
+
+    def _write_db(self, db: JSONs, version: Optional[str]) -> str:
+        """
+        Try to write portal integrations database to S3 and update version.
+        Update is rejected with `VersionConflict` if `version` is not current.
+        :param db: the DB to be written to S3.
+        :param version: the version of the DB this write is intended to replace.
+        :return: version of the newly written DB.
+        """
+        json_bytes = json.dumps(db).encode()
+        response = self.client.put_object(Bucket=config.portal_db_bucket_name,
+                                          Key=config.portal_db_object_key,
+                                          Body=json_bytes,
+                                          ContentType='application/json')
+        new_version = response['VersionId']
+        try:
+            self.version_service.put(self._db_url, version, new_version)
+        except VersionConflict:
+            # Operation was performed on outdated DB and now erroneously exists
+            # in S3.
+            self._delete_db(new_version)
+            raise
+        return new_version
+
+    def _delete_db(self, version: str) -> None:
+        """
+        Delete the specified version of the portal integrations DB from S3.
+        """
+        self.client.delete_object(Bucket=config.portal_db_bucket_name,
```

Comment:
I have seen it happen that this raises NoSuchKey, too. We could either retry in a small loop here or catch, log and ignore.

Reply:
You raise a good point. The only purpose of the `_delete_db` method is to resolve version conflicts in `_write_db`, but it could be that the object to be written won't actually uploaded by the time the version conflict is detected and `_delete_db` is called.

As it stands, the call to `_delete_db` isn't necessary per se, since the copy we delete never has its version ID set as current and doesn't affect further operations. Perhaps we need a stronger strategy in general for removing unneeded versions. For now I think I'll just log and ignore.

Reply:
OK.